torch/csrc/jit/ir.cpp

```diff
@@ -251,6 +255,330 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+    std::ostream & out,
+    const Node* node,
+    const size_t level
+  ) {
+    indent(out, level);
+    const auto& returns = node->inputs();
+    if (returns.size() > 0) {
+      out << "return ";
+      std::string delimiter = "";
+      if (returns.size() > 1) {
+        printValueList(out, returns);
+      } else {
+        printValue(out, returns[0]);
+      }
+      out << std::endl;
+    }
+    return out;
+  }
+
+  std::ostream& printBlock(
+    std::ostream & out,
+    const Block* root,
+    const size_t level
+  ) {
+    for (const auto* node : root->nodes()) {
+      printNode(out, node, level);
+    }
+
+    printNode(out, root->return_node(), level);
+
+    return out;
+  }
+
+  inline bool isNameUnique(std::string& name, const Value* val) const {
+    auto generated_name_value = generated_names_.find(name);
+    if (generated_name_value != generated_names_.end() &&
+        generated_name_value->second != val) {
+      // Found a generated name match, check that it's for a different value
+      return false;
+    }
+    return graph_.uniqueNames().find(name) == graph_.uniqueNames().end();
+  }
+
+  std::ostream& printValue(
+    std::ostream & out,
+    const Value* val
+  ) {
+    const auto node = val->node();
+    if (node->kind() == prim::Constant) {
+      printAttributeValue(out, node->attributeNames()[0], node);
+      return out;
+    }
+
+    auto name_source = val;
+
+    auto aliased_name = aliases_.find(val);
+    if (aliased_name != aliases_.end()) {
+      name_source = aliased_name->second;
+    }
+
+    auto name = name_source->uniqueName();
+
+    bool using_generated_name = false;
+    if (isdigit(name.at(0))) {
+      std::stringstream ss;
+      ss << "t" << name;
+      name = ss.str();
+      using_generated_name = true;
+    } else if (name.find_last_of('.') != std::string::npos) {
+      // Make unique name a valid variable name (e.g. a.1 -> a1)
+      name.erase(std::remove(name.begin(), name.end(), '.'), name.end());
+      using_generated_name = true;
+    }
+
+    if (using_generated_name) {
+      // Make sure name is unique
+      size_t suffix = 0;
+      while (!isNameUnique(name, name_source)) {
+        std::stringstream ss;
+        ss << name << suffix;
+        name = ss.str();
+        ++suffix;
+      }
+
+      // These names aren't in the Graph's list of names but we still need to
+      // make sure there are no name conflicts
+      generated_names_[name] = name_source;
+    }
+
+    out << name;
+    return out;
+  }
+
+public:
+  PrettyPrintPass(Graph& graph) : graph_(graph) {}
+
+  std::ostream& run(std::ostream & out) {
+    out << "def script";
+    const Node* params = graph_.block()->param_node();
+    printValueList(out, params->outputs());
+    out << ":" << std::endl;
+
+    // Print body
+    printBlock(out, graph_.block(), 1);
+
+    printReturn(out, graph_.block()->return_node(), 1);
+
+    return out;
+  }
+};
+
+std::ostream& Graph::prettyPrint(std::ostream & out) {
```

Comment:
Quick request: can we add a `dumpPretty()` method to `Graph` that invokes this function using `std::out`? That would make it easier to use this function from a debugger.